### PR TITLE
Transfer search

### PR DESCRIPTION
**Pull request by matprst**

### CODE REVIEW

**Review by rahelbeloch — APPROVED** [review]
Looks good from my side.

File: models/model.py [hunk 1]
```diff
@@ -115,17 +129,22 @@ def __init__(self, optimizer, warmup, max_iters):
         super().__init__(optimizer)
 
     def get_lr(self):
-        lr_factor = self.get_lr_factor(epoch=self.last_epoch)
+        lr_factor = self.get_lr_factor(current_step=self.last_epoch)
```
Comment (Matyasch) [inline models/model.py]: So why do we need to pass this as an argument? We could just access it with `self` in `get_lr_factor`

File: models/model.py [hunk 2]
```diff
@@ -54,7 +60,16 @@ def __init__(self, model_hparams, optimizer_hparams):
         self.lr_scheduler = None
 
     def configure_optimizers(self):
-        optimizer = AdamW(self.parameters(), lr=self.hparams.optimizer_hparams['lr'],
+        params = list(self.named_parameters())
+
+        def is_encoder(n): return n.startswith('model')
```
Comment (Matyasch) [inline models/model.py:74]: Does this also work for both GNN architectures?

**Review by Matyasch — COMMENTED** [review]
I'm not too familiar with Roberta but seems mostly good

File: models/model.py (hunk 2 above)
Comment (fvlieshout) [reply models/model.py:74]: Curious about this indeed

File: models/roberta_encoder.py [hunk 3]
```diff
@@ -12,9 +12,15 @@ def __init__(self):
         # this model is in eval per default, we want to fine-tune it but only the top layers
         self.model.train()
 
-        # only freezing the encoder parameters / weights of the head layers
-        for param in self.model.base_model.parameters():
-            param.requires_grad = False
+        # freeze half if we are doing hyper parameter search
```
Comment (fvlieshout) [inline models/roberta_encoder.py:15]: Shouldn't you always freeze half? When doing transfer learning we also want to still train the other half of the model right?

**Review by fvlieshout — COMMENTED** [review]
Still some questions. Otherwise, looks good!

File: models/model.py [hunk 4]
```diff
@@ -54,7 +60,16 @@ def __init__(self, model_hparams, optimizer_hparams):
...
+
+        grouped_parameters = [
+            {'params': [p for n, p in params if is_encoder(n)], 'lr': self.hparams.optimizer_hparams['lr']},
+            {'params': [p for n, p in params if not is_encoder(n)], 'lr': self.hparams.optimizer_hparams['lr'] * 100}
```
Comment (fvlieshout) [inline models/model.py]: Shouldn't we be able to specify the different learning rates ourselves? Why is it times 100?

File: models/roberta_encoder.py (hunk 3 above)
Comment (matprst) [reply models/roberta_encoder.py:15]: During hyperparameter search we freeze half of the layers so it's faster. When doing the final transfer learning we finetune all the layers.

File: models/model.py (hunk 4 above)
Comment (matprst) [reply models/model.py]: Ideally yes, but because we already started training without having 2 different parameters for the 2 learning rates, it would mean we have to redo the hyperparameter search with the 2 learning rate. It also reduce length of the hyperparameter search (for roberta each combination of parameters takes between 1-3h).
times 100 beause the classifier layers are not trained so the update steps should be larger. Why 100 specifically, it's arbitrary. I've seen some implementation using that so I went with that.

File: models/model.py (hunk 1 above)
Comment (matprst) [reply models/model.py]: Not sure. Maybe @RedHilarious knows?

File: models/model.py (hunk 2 above)
Comment (matprst) [reply models/model.py:74]: Good point. I haven't tested it with the new changes you made. But this part is generic. All the encoders (roberta, and both gnn) are stored in ```self.model```, so there layer name will start with "model".

**Comment by matprst** [comment]
@Matyasch @fvlieshout I answered your comments. I will add a 2nd learning rate and weight decay, will push the new changes in a bit

File: models/model.py (hunk 2 above)
Comment (Matyasch) [reply models/model.py:74]: Yea it makes sense now, I thought 'model' might be some magic keyword in Roberta

**Comment by matprst** [comment]
Done, let me know what you think.

File: models/model.py [hunk 5]
```diff
@@ -64,13 +64,22 @@ def configure_optimizers(self):
...
+                'params': [p for n, p in params if not is_encoder(n)], 
+                'lr': self.hparams.optimizer_hparams['lr_cl'],
+                'weight_decay': self.hparams.optimizer_hparams["weight_decay_cl"]
+            }
         ]
 
-        optimizer = AdamW(grouped_parameters, lr=self.hparams.optimizer_hparams['lr'],
-                          weight_decay=self.hparams.optimizer_hparams['weight_decay'])
+        optimizer = AdamW(grouped_parameters, lr=self.hparams.optimizer_hparams['lr_enc'],
+                          weight_decay=self.hparams.optimizer_hparams['weight_decay_enc'])
```
Comment (Matyasch) [inline models/model.py]: `lr` and `weight_decay` are not needed here anymore right?

File: train.py [hunk 6]
```diff
@@ -23,17 +23,17 @@
 SUPPORTED_DATASETS = ['R8', 'R52', 'AGNews', 'IMDb']
 
 
-def train(model_name, seed, epochs, patience, b_size, l_rate, w_decay, warmup, cf_hidden_dim, data_name,
-          resume, gnn_layer_name, transfer, h_search):
+def train(model_name, seed, epochs, patience, b_size, l_rate_enc, l_rate_cl, w_decay_enc, w_decay_cl, warmup, cf_hidden_dim, data_name,
+          resume, gnn_layer_name, transfer, h_search, l_rate2):
```
Comment (Matyasch) [inline train.py]: What is `l_rate2`? Maybe used while testing?

File: models/model.py (hunk 4 above)
Comment (fvlieshout) [reply models/model.py]: But you started training just on the train task right? We didn't need an extra learning rate there. I agree that it complicates the hyperparameter search so maybe we could just go with one of which we think it will work otherwise we won't have any results tomorrow for Roberta. (Training on GNN takes only 2 minutes). If you saw this somewhere, I guess this should be fine and I would opt for using that as well. Not sure what @RedHilarious and @Matyasch think about this?

File: models/roberta_encoder.py (hunk 3 above)
Comment (fvlieshout) [reply models/roberta_encoder.py:15]: Oh really? I thought Phillip recommended only training half of the layers

File: models/model.py (hunk 2 above)
Comment (fvlieshout) [reply models/model.py:74]: I think @RedHilarious actually renamed this file so if that impacts the model name, we should reverse that renaming as we already trained the models with from the models.py file

Comment (fvlieshout) [reply models/model.py:74]: Oh sorry you wrote self.model. I read model.py somehow... I guess then the renaming doesn't matter

File: train.py [hunk 7]
```diff
@@ -160,9 +162,14 @@ def initialize_trainer(epochs, patience, model_name, l_rate, weight_decay, warmu
     parser.add_argument('--epochs', dest='epochs', type=int, default=50)
     parser.add_argument('--patience', dest='patience', type=int, default=10)
     parser.add_argument('--batch-size', dest='batch_size', type=int, default=64)
-    parser.add_argument('--lr', dest='l_rate', type=float, default=0.01)
-    parser.add_argument("--w-decay", dest='w_decay', type=float, default=2e-3,
-                        help="Weight decay for L2 regularization of optimizer AdamW")
+    parser.add_argument('--lr-enc', dest='l_rate_enc', type=float, default=0.01,
+                        help="Encoder learning rate.")
+    parser.add_argument('--lr-cl', dest='l_rate_cl', type=float, default=0.01,
```
Comment (Matyasch) [inline train.py]: So this might be ugly to implement but would it be possible to have the default values as -1 and then later check that if the value is <0 then just make it equal to the encoder params? Just so if we don't want to separate it we can do it easily

File: models/model.py (hunk 2 above)
Comment (matprst) [reply models/model.py:74]: The string "model" is linked to the named of the parameter ```self.model``` (if the parameter was named ```encoder```), then the string would need to be "encoder". In this PR (and in main) that's the correct name. What changes you are referring to @fvlieshout ?

File: models/roberta_encoder.py (hunk 3 above)
Comment (matprst) [reply models/roberta_encoder.py:15]: As far as I understood his comment was for the hyperparameter search, and then finetune the whole model once the best configuration as been found (I could be wrong)

File: train.py [hunk 8]
```diff
@@ -23,37 +23,40 @@
 SUPPORTED_DATASETS = ['R8', 'R52', 'AGNews', 'IMDb']
 
 
-def train(model_name, seed, epochs, patience, b_size, l_rate, w_decay, warmup, max_iters, cf_hidden_dim, data_name,
-          resume, gnn_layer_name, transfer):
+def train(model_name, seed, epochs, patience, b_size, l_rate_enc, l_rate_cl, w_decay_enc, w_decay_cl, warmup, cf_hidden_dim, data_name,
+          resume, gnn_layer_name, transfer, h_search, l_rate2):
     os.makedirs(LOG_PATH, exist_ok=True)
```
Comment (fvlieshout) [inline train.py:28]: What's l_rate2?

File: train.py [hunk 9]
```diff
@@ -191,13 +198,16 @@ def initialize_trainer(epochs, patience, model_name, l_rate, weight_decay, warmu
...
         warmup=params["warmup"],
-        max_iters=params["max_iters"],
         cf_hidden_dim=params["cf_hidden_dim"],
         data_name=params["dataset"],
         resume=params["resume"],
         gnn_layer_name=params["gnn_layer_name"],
-        transfer=params["transfer"]
+        transfer=params["transfer"],
+        h_search=params["h_search"],
+        l_rate2 = 42,
```
Comment (fvlieshout) [inline train.py]: Here again: what's l_rate2?

**Review by fvlieshout — APPROVED** [review]
I think it's good like this. We can set the lr of the classifier to be 100 times the lr for the encoder manually if we decide that that's the way to go but now we also have the freedom to do hyperparameter search on it if we want.
Only one question about l_rate2

File: models/model.py (hunk 2 above)
Comment (fvlieshout) [reply models/model.py:74]: In the other PR from Rahel, she renamed model.py to document_classifier.py (or something similar). I somehow read that the naming of the python file would influence the names of the model layers haha. But nevermind, I just misread!

File: models/roberta_encoder.py (hunk 3 above)
Comment (fvlieshout) [reply models/roberta_encoder.py:15]: Yes I now reread the assignment and you're correct

**Comment by matprst** [comment]
I added a default value of -1 for the classifier lr and weight decay as @Matyasch  suggested. If they are not specified they will be set to the same value as the encoder. 
Sorry about ```l_rate2``` I'm tired ...

**Review by Matyasch — APPROVED** [review]
Perfect

File: models/model.py (hunk 1 above)
Comment (rahelbeloch) [reply models/model.py]: I will adapt it! I just took the Phillip implementation from his tutorial and did not adapt it further.

File: models/model.py (hunk 4 above)
Comment (rahelbeloch) [reply models/model.py]: So I did not think that the * 100 is arbitrary. In the first trainings I did, I unfortunately froze all roberta and only trained the classifier. Here a lr of `5e-3/4` did work well. Now with Roberta we found that a lr of `5e-6`works well, so multiplying it with 100, we end up at the ideal LR for the classifier. That is my reasoning behind it.

We started training on Task 1 and yes, we also used 2 learning rates here (and I also tuned both), the original roberta one and then *100 for the classifier. 

So I think we know which LR works well for the classifier, so the only thing we still need to find out is which LR works good for the GNN.
